### PR TITLE
core: Add `abi` module

### PR DESCRIPTION
**Pull request by moCello**
Move the `abi` module from `rusk-abi` to `core` under a feature flag.

The structs `Metadata` and `Query` are now also under `dusk_core::abi`.
As are the re-exports from `piecrust-uplink`.

The backwards compatibility of these changes has been confirmed with the help of @HDauven and @miloszm

### CODE REVIEW

File: rusk/src/lib/node/rusk.rs [hunk 1]
```diff
@@ -9,33 +9,31 @@ use std::sync::{mpsc, Arc, LazyLock};
...
-    stake::{Reward, RewardReason, StakeData, STAKE_CONTRACT},
-    transfer::{
-        data::ContractDeploy, moonlight::AccountData,
-        Transaction as ProtocolTransaction, TRANSFER_CONTRACT,
-    },
-    BlsScalar, ContractError, Dusk, Event,
+use dusk_core::abi::{ContractError, Event};
+use dusk_core::signatures::bls::PublicKey as BlsPublicKey;
+use dusk_core::stake::StakeKeys;
+use dusk_core::stake::{Reward, RewardReason, StakeData, STAKE_CONTRACT};
```
Comment (HDauven) [inline rusk/src/lib/node/rusk.rs:22]:
```suggestion
use dusk_core::stake::{Reward, RewardReason, StakeData, StakeKeys, STAKE_CONTRACT};
```

File: rusk/src/lib/node/rusk.rs [hunk 2]
```diff
@@ -9,33 +9,31 @@ use std::sync::{mpsc, Arc, LazyLock};
...
-    signatures::bls::PublicKey as BlsPublicKey,
-    stake::{Reward, RewardReason, StakeData, STAKE_CONTRACT},
-    transfer::{
-        data::ContractDeploy, moonlight::AccountData,
-        Transaction as ProtocolTransaction, TRANSFER_CONTRACT,
-    },
-    BlsScalar, ContractError, Dusk, Event,
+use dusk_core::abi::{ContractError, Event};
+use dusk_core::signatures::bls::PublicKey as BlsPublicKey;
+use dusk_core::stake::StakeKeys;
```
Comment (HDauven) [inline rusk/src/lib/node/rusk.rs:21]:
```suggestion
```

File: rusk/src/lib/node/rusk.rs [hunk 3]
```diff
@@ -9,33 +9,31 @@ use std::sync::{mpsc, Arc, LazyLock};
...
+use dusk_core::stake::{Reward, RewardReason, StakeData, STAKE_CONTRACT};
+use dusk_core::transfer::PANIC_NONCE_NOT_READY;
```
Comment (HDauven) [inline rusk/src/lib/node/rusk.rs:23]:
```suggestion
```

File: rusk/src/lib/node/rusk.rs [hunk 4]
```diff
@@ -9,33 +9,31 @@ use std::sync::{mpsc, Arc, LazyLock};
...
+use dusk_core::transfer::{
+    data::ContractDeploy, moonlight::AccountData,
```
Comment (HDauven) [inline rusk/src/lib/node/rusk.rs:25]:
```suggestion
    data::ContractDeploy, moonlight::AccountData, PANIC_NONCE_NOT_READY,
```

File: core/Cargo.toml [hunk 5]
```diff
@@ -37,6 +37,12 @@ parallel = [
     "bls12_381-bls/parallel", 
 ]
 
+# These are the features available for when one wishes to use `rusk-abi` as a
```
Comment (HDauven) [inline core/Cargo.toml:40]:
```suggestion
# These are the features available for when one wishes to use `rusk-abi` in a
```

File: core/src/abi.rs [hunk 6]
```diff
@@ -0,0 +1,194 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+//! The application binary interface for the Dusk-network.
```
Comment (HDauven) [inline core/src/abi.rs:7]:
```suggestion
//! The application Binary Interface (ABI) for the Dusk network.
```

**Review by HDauven — APPROVED** [review]
Very nice, LGTM!

A couple of tiny nits, I'll leave it up to you what to do with them

File: core/Cargo.toml (hunk 5 above)
Comment (moCello) [reply core/Cargo.toml:40]: Actually we also need to scratch the `rusk-abi` part there.

**Review by HDauven — APPROVED** [review]

**Pull request merged**